tmva/tmva/inc/TMVA/DNN/RMSProp.h

```diff
@@ -0,0 +1,247 @@
+ *      CERN, Switzerland                                                         *
+ *      U. of Victoria, Canada                                                    *
+ *      MPI-K Heidelberg, Germany                                                 *
+ *      U. of Bonn, Germany                                                       *
+ *                                                                                *
+ * Redistribution and use in source and binary forms, with or without             *
+ * modification, are permitted according to the terms listed in LICENSE           *
+ * (http://tmva.sourceforge.net/LICENSE)                                          *
+ **********************************************************************************/
+
+#ifndef TMVA_DNN_RMSPROP
+#define TMVA_DNN_RMSPROP
+
+#include "TMatrix.h"
+#include "TMVA/DNN/Optimizer.h"
+#include "TMVA/DNN/Functions.h"
+
+namespace TMVA {
+namespace DNN {
+
+/** \class TRMSProp
+ *  RMSProp Optimizer class
+ * 
+ *  This class represents the RMSProp Optimizer with options for applying momentum.
+ */
+template <typename Architecture_t, typename Layer_t = VGeneralLayer<Architecture_t>,
+          typename DeepNet_t = TDeepNet<Architecture_t, Layer_t>>
+class TRMSProp : public VOptimizer<Architecture_t, Layer_t, DeepNet_t> {
+public:
+   using Matrix_t = typename Architecture_t::Matrix_t;
+   using Scalar_t = typename Architecture_t::Scalar_t;
+
+protected:
+   Scalar_t fMomentum; ///< The momentum used for training.
+   Scalar_t fRho; ///< The Rho constant used by the optimizer.
+   Scalar_t fEpsilon; ///< The Smoothing term used to avoid division by zero.
+   std::vector<std::vector<Matrix_t>>
+      fPastSquaredWeightGradients; ///< The sum of the square of the past weight gradients associated with the deep net.
+   std::vector<std::vector<Matrix_t>>
+      fPastSquaredBiasGradients; ///< The sum of the square of the past bias gradients associated with the deep net.
+
+   std::vector<std::vector<Matrix_t>> fWeightUpdates; ///< The accumulation of the past Weights for performing updates.
+   std::vector<std::vector<Matrix_t>> fBiasUpdates; ///< The accumulation of the past Biases for performing updates.
+
+   /*! Update the weights, given the current weight gradients. */
+   void UpdateWeights(size_t layerIndex, std::vector<Matrix_t> &weights, const std::vector<Matrix_t> &weightGradients);
+
+   /*! Update the biases, given the current bias gradients. */
+   void UpdateBiases(size_t layerIndex, std::vector<Matrix_t> &biases, const std::vector<Matrix_t> &biasGradients);
+
+public:
+   /*! Constructor. */
+   TRMSProp(Scalar_t learningRate, DeepNet_t &deepNet, Scalar_t momentum, Scalar_t rho = 0.9, Scalar_t epsilon = 1e-7);
+
+   /*! Destructor. */
+   ~TRMSProp() = default;
+
+   /*! Getters */
+   Scalar_t GetMomentum() const { return fMomentum; }
+   Scalar_t GetRho() const { return fRho; }
+   Scalar_t GetEpsilon() const { return fEpsilon; }
+
+   /*! Temp Print */
```

Comment:
To be removed?

Reply:
Yeah Yeah I need to remove that print. I declared only for debugging.